magellan-library/src/main/java/com/wealthfront/magellan/navigation/LinearNavigator.kt

```diff
@@ -0,0 +1,187 @@
+
+  private var ghostView: View? = null
+  private var containerView: ScreenContainer? = null
+
+  private val currentNavigable: Navigable?
+    get() = backStack.peek()?.navigable
+
+  private val context: Context?
+    get() = currentState.context
+
+  override fun onCreate(context: Context) {
+    currentNavigable?.let {
+      lifecycleStateMachine.transitionBetweenLifecycleStates(it, Destroyed, Created(context))
+    }
+    lifecycleStateMachine.transitionBetweenLifecycleStates(
+      backStack.navigables(),
+      Destroyed,
+      Created(context))
+  }
+
+  override fun onShow(context: Context) {
+    containerView = container()
+    currentNavigable?.let {
+      containerView!!.addView(it.view!!)
+    }
+  }
+
+  override fun onDestroy(context: Context) {
+    currentNavigable?.let {
+      lifecycleStateMachine.transitionBetweenLifecycleStates(it, Created(context), Destroyed)
+    }
+    lifecycleStateMachine.transitionBetweenLifecycleStates(
+      backStack.navigables(),
+      Created(context),
+      Destroyed)
+    containerView = null
+  }
+
+  fun goTo(nextNavigable: Navigable) {
+    navigateTo(nextNavigable, GO)
+  }
+
+  fun show(nextNavigable: Navigable) {
+    navigateTo(nextNavigable, SHOW)
+  }
+
+  fun replaceAndGo(nextNavigable: Navigable) {
+    replace(nextNavigable, GO)
+  }
+
+  fun replaceAndShow(nextNavigable: Navigable) {
+    replace(nextNavigable, SHOW)
+  }
+
+  private fun replace(nextNavigable: Navigable, navType: NavigationType) {
+    navigate(FORWARD) {
+      backStack.pop()
+      backStack.push(NavigationEvent(nextNavigable, navType))
+    }
+  }
+
+  private fun navigateTo(nextNavigable: Navigable, navType: NavigationType) {
+    navigate(FORWARD) { backStack ->
+      backStack.push(NavigationEvent(nextNavigable, navType))
+    }
+  }
+
+  private fun navigateBack() {
+    navigate(BACKWARD) { backstack ->
+      backstack.pop()
+    }
+  }
+
+  fun navigate(
+    direction: Direction,
+    backstackOperation: (Deque<NavigationEvent>) -> Unit
+  ) {
+    containerView?.setInterceptTouchEvents(true)
+    val from = hideCurrentNavigable(direction)
+    backstackOperation.invoke(backStack)
+    val to = showCurrentNavigable(direction)
+    animateAndRemove(from, to, direction)
+  }
+
+  private fun animateAndRemove(
+    from: View?,
+    to: View?,
+    direction: Direction
+  ) {
+    ghostView = from
+    val navEvent = backStack.peek()!!
+    val transition = DefaultTransition()
+    to?.whenMeasured {
+      transition.animate(from, to, navEvent.navigationType, direction) {
+        if (context != null) {
+          containerView!!.removeView(from)
+          if (from == ghostView) {
+            // Only clear the ghost if it's the same as the view we just removed
+            ghostView = null
+          }
+          currentNavigable!!.transitionFinished()
+          containerView!!.setInterceptTouchEvents(false)
+        }
+      }
```

Comment:
I think we need a `currentNavigable.transitionStarted()` in here somewhere

Reply:
Moved it near the transition start.